core/lib/zksync_core/src/basic_witness_input_producer/mod.rs

```diff
@@ -0,0 +1,190 @@
+use multivm::interface::L2BlockEnv;
+use tokio::runtime::Handle;
+use tokio::task::JoinHandle;
+
+mod metrics;
+mod vm_interactions;
+
+use metrics::BASIC_WITNESS_INPUT_PRODUCER_METRICS;
+use vm_interactions::{create_vm, execute_tx};
+use zksync_types::block::MiniblockExecutionMode;
+
+/// Component that extracts all data (from DB) necessary to run a Basic Witness Generator.
+/// This component will upload Witness Inputs to the object store.
+/// This allows Witness Generator workflow (that needs only Basic Witness Generator Inputs)
+/// to be run only using the object store information, having no other external dependency.
+#[derive(Debug)]
+pub struct BasicWitnessInputProducer {
+    connection_pool: ConnectionPool,
+    l2_chain_id: L2ChainId,
+    object_store: Arc<dyn ObjectStore>,
+}
+
+impl BasicWitnessInputProducer {
+    pub async fn new(
+        connection_pool: ConnectionPool,
+        store_factory: &ObjectStoreFactory,
+        l2_chain_id: L2ChainId,
+    ) -> anyhow::Result<Self> {
+        Ok(BasicWitnessInputProducer {
+            connection_pool,
+            object_store: store_factory.create_store().await.into(),
+            l2_chain_id,
+        })
+    }
+
+    fn process_job_impl(
+        rt_handle: Handle,
+        l1_batch_number: L1BatchNumber,
+        started_at: Instant,
+        connection_pool: ConnectionPool,
+        l2_chain_id: L2ChainId,
+    ) -> anyhow::Result<WitnessBlockState> {
+        let mut connection = rt_handle
+            .block_on(connection_pool.access_storage())
+            .context("failed to get connection for BasicWitnessInputProducer")?;
+
+        let miniblocks_execution_data = rt_handle.block_on(
+            connection
+                .transactions_dal()
+                .get_miniblocks_to_execute_for(MiniblockExecutionMode::L1Batch(l1_batch_number)),
+        )?;
+
+        let (mut vm, storage_view) =
+            create_vm(rt_handle.clone(), l1_batch_number, connection, l2_chain_id)
+                .context("failed to create vm for BasicWitnessInputProducer")?;
+
+        tracing::info!("Started execution of l1_batch: {l1_batch_number:?}");
+        let mut next_miniblocks_data = miniblocks_execution_data.iter().skip(1);
+        for miniblock_execution_data in miniblocks_execution_data.iter() {
+            tracing::debug!(
+                "Started execution of miniblock: {:?}",
+                miniblock_execution_data.number
+            );
+            for tx in &miniblock_execution_data.txs {
+                tracing::trace!("Started execution of tx: {tx:?}");
+                execute_tx(tx, &mut vm)
+                    .context("failed to execute transaction in BasicWitnessInputProducer")?;
+                tracing::trace!("Finished execution of tx: {tx:?}");
+            }
+            if let Some(next_miniblock_data) = next_miniblocks_data.next() {
+                vm.start_new_l2_block(L2BlockEnv::from_miniblock_data(next_miniblock_data));
```

Comment:
We're walking in circles now 😅  It was [my suggestion](https://github.com/matter-labs/zksync-era/pull/156#discussion_r1368217091) to use iterator here.

I'd argue that using indexing is not idiomatic in Rust.

Reply:
![Screenshot 2023-10-24 at 09 13 29](https://github.com/matter-labs/zksync-era/assets/1053184/820582bc-4807-44de-8420-082584d2b86e)

Pasting here for convinience.

Definitely not a big deal and also have no idea what's more idiomatic - but comparing the two versions for me indexes definitely look easier to follow (that mutable variable disgusts me haha also I'm looking for additional usages since the variable definition is not in the same line as usage/rewrite).

Will ask @slowli what he things - more because of curiousity than anything else.

@EmilLuta sorry for back and forth

Reply:
definitely doesn't feel more functional for me - we maintain a mutable variable which is introduce outside of the scope loop. That said, none of the two are functional - and it cannot be achieved here due to the VM interface

Reply:
I don't quite agree. This is just a manual implementation of [`Itertools::zip_longest`](https://docs.rs/itertools/latest/itertools/trait.Itertools.html#method.zip_longest). I wanted to suggest using it, but I thought that the API there isn't really good (introduces a separate type as an item, plus using it would require a slightly odd pattern matching given that we know beforehand which iterator is longer).
The mutable variable is an iterator that is meant to be consumed (and which is created under the hood in every `for` loop in fact); the produced items are still immutable.

This _is_ a functional approach, just the Rust dialect of it.

Reply:
I was going to suggest using `slice::windows()` 🙃 I don't quite like in the current approach that the "next item" iterator is iterated manually, rather than together with the "current item" iterator. It could be possible to redefined the "next item" iterator as something like

```rust
let next_miniblocks_data = 
    miniblocks_execution_data.iter().skip(1).map(Some).chain([None]);
```

...and then zip it with the "current item" iterator, right?

Reply:
I'm not against it -- the construction feels a bit harder to parse mentally IMHO, but the loop will look nicer indeed.